File: frontend/src/app/shared/services/lead-crm.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

// ===========================
// Interfaces
// ===========================

export interface Lead {
    id: number;
    tenantId: number;
    name: string;
    email?: string;
    phone?: string;
    country?: string;
    source: string;
    status: string;
    priority: string;
    interestedIn?: string;
    travelDateFrom?: string;
    travelDateTo?: string;
    pax?: number;
    budget?: number;
    currency: string;
    assignedTo?: string;
    notes?: string;
    createdAt: string;
    updatedAt: string;
    convertedAt?: string;
    proposalId?: number;
    followUpCount: number;
    lastFollowUpDate?: string;
}

export interface CreateLeadRequest {
    name: string;
    email?: string;
    phone?: string;
    country?: string;
    source: string;
    priority: string;
    interestedIn?: string;
    travelDateFrom?: string;
    travelDateTo?: string;
    pax?: number;
    budget?: number;
    currency: string;
    assignedTo?: string;
    notes?: string;
}

export interface FollowUp {
    id: number;
    leadId: number;
    type: string;
    message: string;
    nextFollowUpDate?: string;
    createdBy?: string;
    createdAt: string;
}

export interface CreateFollowUpRequest {
    leadId: number;
    type: string;
    message: string;
    nextFollowUpDate?: string;
}

export interface Quotation {
    id: number;
    tenantId: number;
    quotationNumber: string;
    leadId?: number;
    leadName?: string;
    proposalId?: number;
    clientName: string;
    clientEmail?: string;
    clientPhone?: string;
    travelDateFrom?: string;
    travelDateTo?: string;
    pax: number;
    subtotal: number;
    discountAmount: number;
    taxAmount: number;
    totalAmount: number;
    currency: string;
    status: string;
    validUntil?: string;
    notes?: string;
    createdAt: string;
    lineItems: QuotationLineItem[];
}

export interface QuotationLineItem {
    id?: number;
    itemType: string;
    description: string;
    quantity: number;
    unitPrice: number;
    totalPrice?: number;
    sortOrder: number;
}

export interface Amendment {
    id: number;
    bookingInstanceId: number;
    bookingReference?: string;
    amendmentType: string;
    description: string;
    oldValue?: string;
    newValue?: string;
    status: string;
    requestedBy?: string;
    approvedBy?: string;
    feeAmount: number;
    createdAt: string;
    resolvedAt?: string;
}

export interface SupplierContract {
    id: number;
    tenantId: number;
    supplierType: string;
    supplierId: number;
    supplierName: string;
    contractNumber?: string;
    startDate: string;
    endDate: string;
    rateType: string;
    baseRate: number;
    currency: string;
    terms?: string;
    status: string;
    createdAt: string;
}

export interface B2BAgent {
    id: number;
    tenantId: number;
    name: string;
    contactPerson?: string;
    email?: string;
    phone?: string;
    country?: string;
    region?: string;
    commissionRate: number;
    creditLimit: number;
    creditBalance: number;
    status: string;
    notes?: string;
    createdAt: string;
    totalBookings: number;
}

export interface AgentPricing {
    id: number;
    agentId: number;
    itineraryId: number;
    itineraryTitle?: string;
    pricePerPerson: number;
    currency: string;
    validFrom?: string;
    validTo?: string;
}

export interface LedgerEntry {
    id: number;
    agentId: number;
    transactionType: string;
    amount: number;
    reference?: string;
    description?: string;
    balanceAfter: number;
    createdAt: string;
}

export interface SupplierPayment {
    id: number;
    tenantId: number;
    supplierType: string;
    supplierId: number;
    supplierName: string;
    bookingInstanceId?: number;
    bookingReference?: string;
    amount: number;
    currency: string;
    paymentMethod?: string;
    paymentDate: string;
    reference?: string;
    status: string;
    notes?: string;
    createdAt: string;
}

export interface ServiceVoucher {
    id: number;
    tenantId: number;
    voucherNumber: string;
    bookingInstanceId?: number;
    bookingReference?: string;
    voucherType: string;
    supplierType?: string;
    supplierId?: number;
    supplierName?: string;
    serviceDate?: string;
    details?: string;
    status: string;
    createdAt: string;
}

export interface RoomingAssignment {
    id: number;
    departureId: number;
    bookingInstanceId: number;
    bookingReference?: string;
    travelerId: number;
    travelerName?: string;
    hotelId?: number;
    hotelName?: string;
    roomType?: string;
    roomNumber?: string;
    checkInDate?: string;
    checkOutDate?: string;
    notes?: string;
}

@Injectable({ providedIn: 'root' })
export class LeadCrmService {
    private baseUrl = `${environment.apiBaseUrl}/LeadCrm`;

    constructor(private http: HttpClient) {}

    // ======================== LEADS ========================
    getLeads(): Observable<Lead[]> {
        return this.http.get<Lead[]>(`${this.baseUrl}/leads`);
    }

    getLead(id: number): Observable<Lead> {
        return this.http.get<Lead>(`${this.baseUrl}/leads/${id}`);
    }

    createLead(data: CreateLeadRequest): Observable<Lead> {
        return this.http.post<Lead>(`${this.baseUrl}/leads`, data);
    }

    updateLead(id: number, data: Partial<Lead>): Observable<Lead> {
        return this.http.put<Lead>(`${this.baseUrl}/leads/${id}`, data);
    }

    deleteLead(id: number): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/leads/${id}`);
    }

    convertLead(leadId: number, proposalId: number): Observable<void> {
        return this.http.post<void>(`${this.baseUrl}/leads/${leadId}/convert/${proposalId}`, {});
    }

    // ======================== FOLLOW-UPS ========================
    getAllFollowUps(): Observable<FollowUp[]> {
        return this.http.get<FollowUp[]>(`${this.baseUrl}/follow-ups`);
    }

    getFollowUps(leadId: number): Observable<FollowUp[]> {
        return this.http.get<FollowUp[]>(`${this.baseUrl}/leads/${leadId}/follow-ups`);
    }

    createFollowUp(data: CreateFollowUpRequest): Observable<FollowUp> {
        return this.http.post<FollowUp>(`${this.baseUrl}/follow-ups`, data);
    }

    // ======================== QUOTATIONS ========================
    getQuotations(): Observable<Quotation[]> {
        return this.http.get<Quotation[]>(`${this.baseUrl}/quotations`);
    }

    getQuotation(id: number): Observable<Quotation> {
        return this.http.get<Quotation>(`${this.baseUrl}/quotations/${id}`);
    }

    createQuotation(data: any): Observable<Quotation> {
        return this.http.post<Quotation>(`${this.baseUrl}/quotations`, data);
    }

    updateQuotationStatus(id: number, status: string): Observable<void> {
        return this.http.put<void>(`${this.baseUrl}/quotations/${id}/status`, { status });
    }

    // ======================== AMENDMENTS ========================
    getAmendments(): Observable<Amendment[]> {
        return this.http.get<Amendment[]>(`${this.baseUrl}/amendments`);
    }

    createAmendment(data: any): Observable<Amendment> {
        return this.http.post<Amendment>(`${this.baseUrl}/amendments`, data);
    }

    updateAmendmentStatus(id: number, status: string): Observable<void> {
        return this.http.put<void>(`${this.baseUrl}/amendments/${id}/status`, { status });
    }

    // ======================== CONTRACTS ========================
    getContracts(): Observable<SupplierContract[]> {
        return this.http.get<SupplierContract[]>(`${this.baseUrl}/contracts`);
    }

    createContract(data: any): Observable<SupplierContract> {
        return this.http.post<SupplierContract>(`${this.baseUrl}/contracts`, data);
    }

    updateContractStatus(id: number, status: string): Observable<void> {
        return this.http.put<void>(`${this.baseUrl}/contracts/${id}/status`, { status });
    }

    // ======================== B2B AGENTS ========================
    getAgents(): Observable<B2BAgent[]> {
        return this.http.get<B2BAgent[]>(`${this.baseUrl}/agents`);
    }

    getAgent(id: number): Observable<B2BAgent> {
        return this.http.get<B2BAgent>(`${this.baseUrl}/agents/${id}`);
    }

    createAgent(data: any): Observable<B2BAgent> {
        return this.http.post<B2BAgent>(`${this.baseUrl}/agents`, data);
    }

    updateAgentStatus(id: number, status: string): Observable<void> {
        return this.http.put<void>(`${this.baseUrl}/agents/${id}/status`, { status });
    }

    // ======================== B2B PRICING ========================
    getAgentPricing(agentId: number): Observable<AgentPricing[]> {
        return this.http.get<AgentPricing[]>(`${this.baseUrl}/agents/${agentId}/pricing`);
    }

    createAgentPricing(data: any): Observable<AgentPricing> {
        return this.http.post<AgentPricing>(`${this.baseUrl}/agent-pricing`, data);
    }

    deleteAgentPricing(id: number): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/agent-pricing/${id}`);
    }

    // ======================== B2B LEDGER ========================
    getAgentLedger(agentId: number): Observable<LedgerEntry[]> {
        return this.http.get<LedgerEntry[]>(`${this.baseUrl}/agents/${agentId}/ledger`);
    }

    createLedgerEntry(data: any): Observable<LedgerEntry> {
        return this.http.post<LedgerEntry>(`${this.baseUrl}/ledger`, data);
    }

    // ======================== SUPPLIER PAYMENTS ========================
    getSupplierPayments(): Observable<SupplierPayment[]> {
        return this.http.get<SupplierPayment[]>(`${this.baseUrl}/supplier-payments`);
    }

    createSupplierPayment(data: any): Observable<SupplierPayment> {
        return this.http.post<SupplierPayment>(`${this.baseUrl}/supplier-payments`, data);
    }

    updateSupplierPaymentStatus(id: number, status: string): Observable<void> {
        return this.http.put<void>(`${this.baseUrl}/supplier-payments/${id}/status`, { status });
    }

    // ======================== VOUCHERS ========================
    getVouchers(): Observable<ServiceVoucher[]> {
        return this.http.get<ServiceVoucher[]>(`${this.baseUrl}/vouchers`);
    }

    createVoucher(data: any): Observable<ServiceVoucher> {
        return this.http.post<ServiceVoucher>(`${this.baseUrl}/vouchers`, data);
    }

    updateVoucherStatus(id: number, status: string): Observable<void> {
        return this.http.put<void>(`${this.baseUrl}/vouchers/${id}/status`, { status });
    }

    // ======================== ROOMING ========================
    getRooming(departureId: number): Observable<RoomingAssignment[]> {
        return this.http.get<RoomingAssignment[]>(`${this.baseUrl}/rooming/${departureId}`);
    }

    createRooming(data: any): Observable<RoomingAssignment> {
        return this.http.post<RoomingAssignment>(`${this.baseUrl}/rooming`, data);
    }

    deleteRooming(id: number): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/rooming/${id}`);
    }
}

File: backend/Bussiness/Services/TourAndTravels/LeadCrmService.cs
using System;
using System.Collections.Generic;
using AutoMapper;
using Domain.Models.TourAndTravels;
using Repository.Interfaces.TourAndTravels;
using static Domain.Models.TourAndTravels.LeadCrm;

namespace Bussiness.Services.TourAndTravels
{
    public interface ILeadCrmService
    {
        // Leads
        LeadResponse CreateLead(CreateLeadRequest request, long tenantId);
        LeadResponse? UpdateLead(long id, UpdateLeadRequest request);
        List<LeadResponse> GetAllLeads(long tenantId);
        LeadResponse? GetLeadById(long id);
        bool DeleteLead(long id);
        bool ConvertLeadToProposal(long leadId, long proposalId);

        // Follow-ups
        FollowUpResponse CreateFollowUp(CreateFollowUpRequest request, string? createdBy);
        List<FollowUpResponse> GetFollowUpsByLead(long leadId);
        List<FollowUpResponse> GetAllFollowUps(long tenantId);

        // Quotations
        QuotationResponse CreateQuotation(CreateQuotationRequest request, long tenantId);
        List<QuotationResponse> GetAllQuotations(long tenantId);
        QuotationResponse? GetQuotationById(long id);
        bool UpdateQuotationStatus(long id, string status);

        // Amendments
        AmendmentResponse CreateAmendment(CreateAmendmentRequest request, long tenantId, string? requestedBy);
        List<AmendmentResponse> GetAllAmendments(long tenantId);
        bool UpdateAmendmentStatus(long id, string status, string? approvedBy);

        // Contracts
        ContractResponse CreateContract(CreateContractRequest request, long tenantId);
        List<ContractResponse> GetAllContracts(long tenantId);
        bool UpdateContractStatus(long id, string status);

        // B2B Agents
        B2BAgentResponse CreateAgent(CreateB2BAgentRequest request, long tenantId);
        List<B2BAgentResponse> GetAllAgents(long tenantId);
        B2BAgentResponse? GetAgentById(long id);
        bool UpdateAgentStatus(long id, string status);

        // B2B Pricing
        AgentPricingResponse CreateAgentPricing(CreateAgentPricingRequest request);
        List<AgentPricingResponse> GetAgentPricing(long agentId);
        bool DeleteAgentPricing(long id);

        // B2B Ledger
        LedgerEntryResponse CreateLedgerEntry(CreateLedgerEntryRequest request);
        List<LedgerEntryResponse> GetAgentLedger(long agentId);

        // Supplier Payments
        SupplierPaymentResponse CreateSupplierPayment(CreateSupplierPaymentRequest request, long tenantId);
        List<SupplierPaymentResponse> GetAllSupplierPayments(long tenantId);
        bool UpdateSupplierPaymentStatus(long id, string status);

        // Vouchers
        VoucherResponse CreateVoucher(CreateVoucherRequest request, long tenantId);
        List<VoucherResponse> GetAllVouchers(long tenantId);
        bool UpdateVoucherStatus(long id, string status);

        // Rooming
        RoomingResponse CreateRooming(CreateRoomingRequest request);
        List<RoomingResponse> GetRoomingByDeparture(long departureId);
        bool DeleteRooming(long id);
    }

    public class LeadCrmService : ILeadCrmService
    {
        private readonly ILeadCrmRepository _repository;
        private readonly IMapper _mapper;

        public LeadCrmService(ILeadCrmRepository repository, IMapper mapper)
        {
            _repository = repository;
            _mapper = mapper;
        }

        // Leads
        public LeadResponse CreateLead(CreateLeadRequest request, long tenantId)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateLeadRequest>(request);
            dto.TenantId = tenantId;
            var result = _repository.CreateLead(dto);
            return _mapper.Map<LeadResponse>(result);
        }

        public LeadResponse? UpdateLead(long id, UpdateLeadRequest request)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.UpdateLeadRequest>(request);
            var result = _repository.UpdateLead(id, dto);
            return result != null ? _mapper.Map<LeadResponse>(result) : null;
        }

        public List<LeadResponse> GetAllLeads(long tenantId)
        {
            var results = _repository.GetAllLeads(tenantId);
            return _mapper.Map<List<LeadResponse>>(results);
        }

        public LeadResponse? GetLeadById(long id)
        {
            var result = _repository.GetLeadById(id);
            return result != null ? _mapper.Map<LeadResponse>(result) : null;
        }

        public bool DeleteLead(long id) => _repository.DeleteLead(id);
        public bool ConvertLeadToProposal(long leadId, long proposalId) => _repository.ConvertLeadToProposal(leadId, proposalId);

        // Follow-ups
        public FollowUpResponse CreateFollowUp(CreateFollowUpRequest request, string? createdBy)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateFollowUpRequest>(request);
            dto.CreatedBy = createdBy;
            var result = _repository.CreateFollowUp(dto);
            return _mapper.Map<FollowUpResponse>(result);
        }

        public List<FollowUpResponse> GetFollowUpsByLead(long leadId)
        {
            return _mapper.Map<List<FollowUpResponse>>(_repository.GetFollowUpsByLead(leadId));
        }

        public List<FollowUpResponse> GetAllFollowUps(long tenantId)
        {
            return _mapper.Map<List<FollowUpResponse>>(_repository.GetAllFollowUps(tenantId));
        }

        // Quotations
        public QuotationResponse CreateQuotation(CreateQuotationRequest request, long tenantId)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateQuotationRequest>(request);
            dto.TenantId = tenantId;
            var result = _repository.CreateQuotation(dto);
            return _mapper.Map<QuotationResponse>(result);
        }

        public List<QuotationResponse> GetAllQuotations(long tenantId) =>
            _mapper.Map<List<QuotationResponse>>(_repository.GetAllQuotations(tenantId));

        public QuotationResponse? GetQuotationById(long id)
        {
            var result = _repository.GetQuotationById(id);
            return result != null ? _mapper.Map<QuotationResponse>(result) : null;
        }

        public bool UpdateQuotationStatus(long id, string status) => _repository.UpdateQuotationStatus(id, status);

        // Amendments
        public AmendmentResponse CreateAmendment(CreateAmendmentRequest request, long tenantId, string? requestedBy)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateAmendmentRequest>(request);
            dto.TenantId = tenantId;
            dto.RequestedBy = requestedBy;
            return _mapper.Map<AmendmentResponse>(_repository.CreateAmendment(dto));
        }

        public List<AmendmentResponse> GetAllAmendments(long tenantId) =>
            _mapper.Map<List<AmendmentResponse>>(_repository.GetAllAmendments(tenantId));

        public bool UpdateAmendmentStatus(long id, string status, string? approvedBy) =>
            _repository.UpdateAmendmentStatus(id, status, approvedBy);

        // Contracts
        public ContractResponse CreateContract(CreateContractRequest request, long tenantId)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateContractRequest>(request);
            dto.TenantId = tenantId;
            return _mapper.Map<ContractResponse>(_repository.CreateContract(dto));
        }

        public List<ContractResponse> GetAllContracts(long tenantId) =>
            _mapper.Map<List<ContractResponse>>(_repository.GetAllContracts(tenantId));

        public bool UpdateContractStatus(long id, string status) => _repository.UpdateContractStatus(id, status);

        // B2B Agents
        public B2BAgentResponse CreateAgent(CreateB2BAgentRequest request, long tenantId)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateB2BAgentRequest>(request);
            dto.TenantId = tenantId;
            return _mapper.Map<B2BAgentResponse>(_repository.CreateAgent(dto));
        }

        public List<B2BAgentResponse> GetAllAgents(long tenantId) =>
            _mapper.Map<List<B2BAgentResponse>>(_repository.GetAllAgents(tenantId));

        public B2BAgentResponse? GetAgentById(long id)
        {
            var result = _repository.GetAgentById(id);
            return result != null ? _mapper.Map<B2BAgentResponse>(result) : null;
        }

        public bool UpdateAgentStatus(long id, string status) => _repository.UpdateAgentStatus(id, status);

        // B2B Pricing
        public AgentPricingResponse CreateAgentPricing(CreateAgentPricingRequest request)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateAgentPricingRequest>(request);
            return _mapper.Map<AgentPricingResponse>(_repository.CreateAgentPricing(dto));
        }

        public List<AgentPricingResponse> GetAgentPricing(long agentId) =>
            _mapper.Map<List<AgentPricingResponse>>(_repository.GetAgentPricing(agentId));

        public bool DeleteAgentPricing(long id) => _repository.DeleteAgentPricing(id);

        // B2B Ledger
        public LedgerEntryResponse CreateLedgerEntry(CreateLedgerEntryRequest request)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateLedgerEntryRequest>(request);
            return _mapper.Map<LedgerEntryResponse>(_repository.CreateLedgerEntry(dto));
        }

        public List<LedgerEntryResponse> GetAgentLedger(long agentId) =>
            _mapper.Map<List<LedgerEntryResponse>>(_repository.GetAgentLedger(agentId));

        // Supplier Payments
        public SupplierPaymentResponse CreateSupplierPayment(CreateSupplierPaymentRequest request, long tenantId)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateSupplierPaymentRequest>(request);
            dto.TenantId = tenantId;
            return _mapper.Map<SupplierPaymentResponse>(_repository.CreateSupplierPayment(dto));
        }

        public List<SupplierPaymentResponse> GetAllSupplierPayments(long tenantId) =>
            _mapper.Map<List<SupplierPaymentResponse>>(_repository.GetAllSupplierPayments(tenantId));

        public bool UpdateSupplierPaymentStatus(long id, string status) =>
            _repository.UpdateSupplierPaymentStatus(id, status);

        // Vouchers
        public VoucherResponse CreateVoucher(CreateVoucherRequest request, long tenantId)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateVoucherRequest>(request);
            dto.TenantId = tenantId;
            return _mapper.Map<VoucherResponse>(_repository.CreateVoucher(dto));
        }

        public List<VoucherResponse> GetAllVouchers(long tenantId) =>
            _mapper.Map<List<VoucherResponse>>(_repository.GetAllVouchers(tenantId));

        public bool UpdateVoucherStatus(long id, string status) => _repository.UpdateVoucherStatus(id, status);

        // Rooming
        public RoomingResponse CreateRooming(CreateRoomingRequest request)
        {
            var dto = _mapper.Map<Repository.DataModels.TourAndTravels.LeadCrmDTO.CreateRoomingRequest>(request);
            return _mapper.Map<RoomingResponse>(_repository.CreateRooming(dto));
        }

        public List<RoomingResponse> GetRoomingByDeparture(long departureId) =>
            _mapper.Map<List<RoomingResponse>>(_repository.GetRoomingByDeparture(departureId));

        public bool DeleteRooming(long id) => _repository.DeleteRooming(id);
    }
}
